python/caffe/layers/detectnet/clustering.py

```diff
@@ -0,0 +1,210 @@
+
+    def reshape(self, bottom, top):
+        n_images = bottom[0].data.shape[0]
+        # Assuming that max booxes per image are MAX_BOXES
+        top[0].reshape(n_images,MAX_BOXES,5)
+
+    def forward(self, bottom, top):
+        bbox = cluster(self,bottom[0].data,bottom[1].data)
+        top[0].data[...] = bbox
+
+    def backward(self, top, propagate_down, bottom):
+        pass
+
+
+class ClusterDetections(caffe.Layer):
+    """
+    * convert network output in grid format to list using group rectangles clustering.
+    * bottom[0] - predicted coverage [ batch_size x grid_sz_x x grid_sz_y x 1 ]
+    * bottom[1] - predicted bbox [batch_size x grid_sz_x x grid_sz_y x 4 (xl, yt, xr, yb)]
+    * top [0] - list of predicted bbox
+        [ batch_size x max_bbox_per_image x 5 (xl, yt, xr, yb, confidence) ]
+
+    Example prototxt definition:
+
+    layer {
+        type: 'Python'
+        name: 'cluster'
+        # det_bbox_list is a batch_size x MAX_BOXES x 5 blob
+        top: 'det_bbox_list'
+        bottom: 'coverage'
+        bottom: 'bbox/regressor'
+        python_param {
+            module: 'caffe.layers.detectnet.clustering'
+            layer: 'ClusterDetections'
+            # parameters - img_size_x, img_size_y, stride,
+            # gridbox_cvg_threshold,gridbox_rect_threshold,gridbox_rect_eps,min_height
+            param_str : '1248,352,16,0.05,1,0.025,22'
+        }
+        include: { phase: TEST }
+    }
+    """
+
+    def setup(self, bottom, top):
+        self.is_groundtruth = False
+        try:
+            plist = self.param_str.split(',')
+            self.image_size_x = int(plist[0])
+            self.image_size_y = int(plist[1])
+            self.stride = int(plist[2])
+            self.gridbox_cvg_threshold = float(plist[3])
+            self.gridbox_rect_thresh = int(plist[4])
+            self.gridbox_rect_eps = float(plist[5])
+            self.min_height = int(plist[6])
+        except ValueError:
+            raise ValueError("Parameter string missing or data type is wrong !")
+
+    def reshape(self, bottom, top):
+        n_images = bottom[0].data.shape[0]
+        # Assuming that max booxes per image are MAX_BOXES
+        top[0].reshape(n_images,MAX_BOXES,5)
+
+    def forward(self, bottom, top):
+        bbox = cluster(self,bottom[0].data,bottom[1].data)
+        top[0].data[...] = bbox
+
+    def backward(self, top, propagate_down, bottom):
+        pass
+
+
+def gridbox_to_boxes(net_cvg, net_boxes, self ):
+    im_sz_x =  self.image_size_x
+    im_sz_y =  self.image_size_y
+    stride =  self.stride
+
+    grid_sz_x = int(im_sz_x / stride)
+    grid_sz_y = int(im_sz_y / stride)
+
+    boxes = []
+    cvgs = []
+
+    cell_width = im_sz_x / grid_sz_x
+    cell_height = im_sz_y / grid_sz_y
+
+    cvg_val = net_cvg[0][0:grid_sz_y][0:grid_sz_x]
+
+    if (self.is_groundtruth):
+        mask=cvg_val>0
+    else:
+        mask=cvg_val>= self.gridbox_cvg_threshold
```

Comment:
Fixed the spacing, but I'm leaving the if/else for readability. One-liners are not always better.